src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs

```diff
@@ -429,36 +429,31 @@ internal abstract partial class SafeDeleteContext : SafeHandle
 
             // Optional output buffer that may need to be freed.
             SafeFreeContextBuffer outFreeContextBuffer = null;
+            Span<IntPtr> ptr = stackalloc IntPtr[3];
             try
             {
-                Span<Interop.SspiCli.SecBuffer> inUnmanagedBuffer = stackalloc Interop.SspiCli.SecBuffer[inSecurityBufferDescriptor.cBuffers];
+                Span<Interop.SspiCli.SecBuffer> inUnmanagedBuffer = stackalloc Interop.SspiCli.SecBuffer[inSecBuffers.Count];
                 inUnmanagedBuffer.Clear();
 
                 fixed (void* inUnmanagedBufferPtr = inUnmanagedBuffer)
-                fixed (void* pinnedToken0 = inSecBuffers.Length > 0 ? inSecBuffers[0].token : null)
-                fixed (void* pinnedToken1 = inSecBuffers.Length > 1 ? inSecBuffers[1].token : null)
-                fixed (void* pinnedToken2 = inSecBuffers.Length > 2 ? inSecBuffers[2].token : null) // pin all buffers, even if null or not used, to avoid needing to allocate GCHandles
+                fixed (void* pinnedToken0 = &inSecBuffers.GetBuffer(0).Token.GetPinnableReference())
+                fixed (void* pinnedToken1 = &inSecBuffers.GetBuffer(1).Token.GetPinnableReference())
+                fixed (void* pinnedToken2 = &inSecBuffers.GetBuffer(2).Token.GetPinnableReference())
                 {
-                    Debug.Assert(inSecBuffers.Length <= 3);
+                    ptr[0] = (IntPtr)pinnedToken0;
+                    ptr[1] = (IntPtr)pinnedToken1;
+                    ptr[2] = (IntPtr)pinnedToken2;
```

Comment:
Why is this span needed at all?  I'm not understanding the purpose.

Reply:
(If you do keep it, please move it to be defined inside of the block for the fixed statements.  It's dangerous to have those pointers in scope after the fixed blocks.

Reply:
I reworked the logic. I removed `inUnmanagedBuffer.Clear()` above and I assign most values instead. If we need to use pinnedTokenX, I do it directly now instead of using temporary storage I can iterate with loop.